### PR TITLE
Upgrade pattern matcher, simplifier, add rules [in progress]

### PR DESCRIPTION
**Pull request by sritchie**

### CODE REVIEW

**Comment by littleredcomputer** [comment]
What is the rationale for implementing new-simplifier? Do you want to be able to import the simplification sequence as it stands in sum?

File: src/sicmutils/rules.cljc [hunk 1]
```diff
@@ -189,70 +235,228 @@
...
+   ))
+
+(def log-extra
+  (ruleset
+   (* (:? n v/integral?) :f1* (log :x) :f2*)
+   =>
+   (* :f1* (log (expt :x :n)) :f2*)
+   ))
+
+(def canonicalize-partials
```
Comment (sritchie) [inline src/sicmutils/rules.cljc:332]: I added this in #286 , remove it here.

File: src/sicmutils/rules.cljc [hunk 2]
```diff
@@ -189,70 +235,228 @@
...
+                                 (eq? (car thing) 'down))))
+                    :else false))])
+    (if (= (length args) 1)
+      (sea access-chain (first args))
+      (sea (rest access-chain)
+           (nth args (first access-chain))))))
 
-(def ^:private flush-obvious-ones
+;; TODO taken from scheme.
+(def canonicalize-partials*
```
Comment (sritchie) [inline src/sicmutils/rules.cljc:371]: whoops, this is what I'd added!

**Comment by breandan** [comment]
It might also be interesting to compare with Tate et al.'s work on [equality saturation](https://www.cs.cornell.edu/~ross/publications/eqsat/). Not sure if there is a JVM implementation of [egg](https://egraphs-good.github.io/), but I saw @erdos has some [experiments in Clojure](https://github.com/erdos/cogent). cc: @mwillsey

**Comment by sritchie** [comment]
Closing in favor of #353 .

**Pull request closed without merging**